### DISCUSSION TITLE
security concerns with dotenv and next.config.js

### DISCUSSION THREAD

**Discussion by LaCocoRoco** [vercel/next.js#40610] [Help] [answered]
To use environment variables in the browser the [NEXT_PUBLIC_](https://nextjs.org/docs/basic-features/environment-variables) prefix is necessary. There is also the possibility to [add variables into the next.config.js](https://nextjs.org/docs/api-reference/next.config.js/environment-variables), which makes the NEXT_PUBLIC_ obsolete (Makes code cleaner). But because other tools also need the .env file, i added the dotenv module into next.config.js (Best of both worlds). Are there any security issues with this?

```javascript
const withPlugins = require('next-compose-plugins');
const nextTranslate = require('next-translate');
const dotenv = require('dotenv-flow');

const configuration = dotenv.config({ silent: true });

const nextConfig = {
  reactStrictMode: true,
  env: configuration.parsed,
};

module.exports = withPlugins([[nextTranslate]], nextConfig);
```

**Comment by icyJoseph** [comment, marked as answer]
Well, at build time Next.js replaces the calls to `process.env.customKey` by the actual value they have. 

Naturally, we'd want to avoid leaking secrets to the client side code. For the most part, function inlining (at build-time), and proper tree-shaking can take care of this. 

Of course, not inlining secrets into components(both at module level, function scope and returned JSX) is also important to keep in mind. 

However, as an application grows, and dependencies become more entangled, and modules are used, with side-effects, both in client and server code, the possibility must be considered seriously.

Take for example:

```ts
// lib/utils.ts
export const add = (a = 0, b = 0) => a + b;

export const logs = () => {
  console.log(process.env.secret);
};
```

I can import this normally on my pages:

```tsx
import type { NextPage } from "next";
import { add } from "../lib/utils";

const Home: NextPage = () => {

  return (
    <div>
      <code>{add(3, 4)}</code>
    </div>
  );
};

export default Home;
```

Then I can search the entire output and in the client side code, and nowhere is my secret to be found.

> Except .next/required-server-files.json which should be pretty alright, as it is not reachable

Now, let's say my utils grew and I want to lazy load them:

```tsx
import type { NextPage } from "next";
import { useState, useEffect } from "react";

const Home: NextPage = () => {
  const [result, setResult] = useState(0);

  useEffect(() => {
    import("../lib/utils").then(({ add }) => setResult(add(3, 4)));
  }, []);

  return (
    <div>
      <code>{result}</code>
    </div>
  );
};

export default Home;
```

I've made, without noticing, a chunk out of `lib/utils` which is downloadable... and if I run my application and look through the chunks... there is it, in the network tab, my secret. A dot-`.env` file loaded secret that doesn't start with `NEXT_PUBLIC` wouldn't be included in the chunk, thus keeping my secret safe.

**Reply by LaCocoRoco** [reply]
Thank you for this beautiful explanation.
This is a very good and import point to consider.
I think this is a good enough reason to roll back.

**Reply by icyJoseph** [reply]
Hi,

There was one thing I had missed here, though it changes nothing, and that is that through magic comments, you can tell webpack to ignore parts of a dynamically imported module, https://webpack.js.org/blog/2020-10-10-webpack-5-release/#general-tree-shaking-improvements.

```ts
 useEffect(() => {
    import(/* webpackExports: ["add"] */ "../lib/utils").then(({ add }) => setResult(add(3, 4)));
  }, []);
```

And that'll make sure I only download `add`, but still, should someone forget the magic comment and that's it, leaked secrets.
